src/classes/CW_CampaignTriggerHandler.cls

```diff
@@ -0,0 +1,308 @@
+    	List<CW_Campaign__c> objects = new List<CW_Campaign__c>();
+    	
+    	for (CW_Campaign__c c :campaigns) {
+			// only include those campaigns that have a roadmap parent
+			if ( c.Roadmap__c!=null ) objects.add(c);
+    	}
+			
+    	/*
+      		First step is to create a context for LREngine, by specifying parent and child objects and
+      		lookup relationship field name
+     	*/
+     	LREngine.Context ctx = new LREngine.Context(
+			CW_Roadmap__c.SobjectType, // parent object
+			CW_Campaign__c.SobjectType,  // child object
+			Schema.SObjectType.CW_Campaign__c.fields.Roadmap__c // relationship field name
+		);
+		
+		/*
+	      Next, one can add multiple rollup fields on the above relationship. 
+	      Here specify 
+	       1. The field to aggregate in child object
+	       2. The field to which aggregated value will be saved in master/parent object
+	       3. The aggregate operation to be done i.e. SUM, AVG, COUNT, MIN/MAX
+     	*/
+     	
+		ctx.add(
+			new LREngine.RollupSummaryField(
+				Schema.SObjectType.CW_Roadmap__c.fields.Points_Used__c,
+				Schema.SObjectType.CW_Campaign__c.fields.Points__c,
+				LREngine.RollupOperation.Sum 
+			)
+		); 
+	
+		/* 
+			Calling rollup method returns in memory master objects with aggregated values in them. 
+			Please note these master records are not persisted back, so that client gets a chance 
+			to post process them after rollup
+		*/ 
+		Sobject[] masters = LREngine.rollUp(ctx, objects);  
+		
+		// Persist the changes in master
+		if (!masters.isEmpty()) update masters;
+    }
+    
+    // Update Roadmap points where campaign previously had a roadmap; remove its campaigns points from the Roadmap (parent)
+    // Key of map is Roadmap Id, Value is list of Campaigns that previously had lookup to Roadmap key
+    public void updateRoadmapWhenCampaignsHaveBeenReparented(Map<Id,List<CW_Campaign__c>> reparentedObjectMap) {
+    	
+    	Map<Id,CW_Roadmap__c> roadmaps = new Map<Id,CW_Roadmap__c>([select Id, Points_Used__c from CW_Roadmap__c where Id in :reparentedObjectMap.keySet()]);
+		List<CW_Roadmap__c> roadmapsToUpdate = new List<CW_Roadmap__c>();
+		
+		for (Id roadmapId :reparentedObjectMap.keySet()) {
+			Double pointsToRemove = 0;
+			List<CW_Campaign__c> camps = reparentedObjectMap.get(roadmapId);
+			for (CW_Campaign__c c :camps) {
+				pointsToRemove += c.Points__c;
+			}
+			CW_Roadmap__c roadmap = roadmaps.get(roadmapId);
+			if (pointsToRemove!=0) {
+				roadmap.Points_Used__c -= pointsToRemove;
+				roadmapsToUpdate.add(roadmap);
+			}
+			
+		}
+		
+		if (!roadmapsToUpdate.isEmpty()) update roadmapsToUpdate;
+    }
+    
+    // update campaign stage on schedules with associated stage on campaign
+    public static void updateChildSchedulesWithCampaignStage(Map<Id,CW_Campaign__c> campaigns) {
+    	
+    	List<CW_Schedule__c> schedules = [select Id, Marketing_Campaign__c, Campaign_Stage__c from CW_Schedule__c where Marketing_Campaign__c in :campaigns.keySet()];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.